Counting linear extensions using gaussian hmc

abureyanahmed opened the pull request:
We have incorporated necessary changes to apply gaussian hmc to count linear extensions. Specifically, we have made the following changes:

Added a new option to use gaussian hmc in the executable file.

Previously the algorithm was unable to compute large volumes; now we are using long double instead of double to fix that.

Changed the scope of a crucial structure random variable to fix a bug.

There were some hardcoded templates, that assumed specific random walk. We have made necessary modifications to support any type of walk.

abureyanahmed committed:
Added gaussian HMC option

Besides using cooling Gaussian, we can now also use gaussian HMC; use "hg" to run the gaussian HMC algorithm.
Also, fixed some bugs, previously the algorithm was unable to compute large volumes; now we are using long double instead of double to fix that.

abureyanahmed committed:
Enhance cooling gaussians algo to use gaussian hmc

There were some hardcoded templates, that assumed specific random walk. We have made necessary modifications to support any type of walk.

abureyanahmed committed:
Fixed a bug in the walk

The _Len variable was a structure variable. Since the structure variable shares memory space, the variable was not getting updated properly. We have changed the scope to update it properly.

abureyanahmed committed:
Enhance cooling gaussians algo to use gaussian hmc

There were some hardcoded templates, that assumed specific random walk. We have made necessary modifications to support any type of walk.

vissarion commented on include/random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp:
```diff
@@ -10,10 +10,17 @@
 
 #include "sampling/sphere.hpp"
 
+#include <iostream>
```
this is only used for debugging, please remove it

vissarion commented on include/random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp:
```diff
@@ -10,10 +10,17 @@
 
+#include <iostream>
 
+typedef long double LNT;
```
this is not needed

vissarion commented on include/random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp:
```diff
@@ -10,10 +10,17 @@
+#include <iostream>
 
+typedef long double LNT;
+typedef double NT;
```
NT is used for the template parameter of Number Type, it is confusing if you define an alias here.

vissarion commented on include/random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp:
```diff
@@ -10,10 +10,17 @@
 
+
+//LNT _Len;
+NT _Len;
```
please do not use global variables, it is error-prone and we avoid it in the whole library.

vissarion commented on include/volume/volume_cooling_gaussians.hpp:
```diff
@@ -22,28 +22,47 @@
 
 template <typename Polytope, typename RandomNumberGenerator>
-struct update_delta<GaussianBallWalk::Walk<Polytope, RandomNumberGenerator>>
+//struct update_delta<GaussianBallWalk::Walk<Polytope, RandomNumberGenerator>>
```
why removing the specialization of `GaussianBallWalk::Walk<Polytope, RandomNumberGenerator>`? 
You can keep both.

vissarion commented on include/volume/volume_cooling_gaussians.hpp:
```diff
@@ -457,7 +513,9 @@ double volume_cooling_gaussians(Polytope const& Pin,
 template
 <
-    typename WalkTypePolicy = GaussianCDHRWalk,
+    typename WalkTypePolicy,
```
why are you removing the default parameter? (same below)

vissarion reviewed (changes requested):
Thanks @abureyanahmed for your first PR in volesti! 
I have some comments inline in the code plus the following general one:
- please remove all comments that are not needed and also debug prints (I understand that helps you when developing / experimenting with the library but it is a pain for reviewers and actually hides the main changes of the contribution)

Also please not that the tests are not passing on CI.

vissarion commented:
> We have incorporated necessary changes to apply gaussian hmc to count linear extensions. Specifically, we have made the following changes:
> 
> Added a new option to use gaussian hmc in the executable file.

You mean the linear extensions example right? This is fine, thanks.

> Previously the algorithm was unable to compute large volumes; now we are using long double instead of double to fix that.

This is using `long double` in linear extensions example, great!
 
> Changed the scope of a crucial structure random variable to fix a bug.

What was the bug about? This is about the global variable that you create? If this is the case then please do not use global variables.
 
> There were some hardcoded templates, that assumed specific random walk. We have made necessary modifications to support any type of walk.

I am not sure I understand this. You mean the default values of the template parameter of the random walk? Why what that a problem?

abureyanahmed committed:
Added gaussian HMC option, no debug print

There were many comments that were just for debugging purposes, I have removed them.

abureyanahmed committed:
Enhance cooling gaussians for gaussian hmc, bugs fixed

There was a subtle issue: the update delta function was taking parameter by value. Hence, the actual change was not taking place. Now we have changed it to a parameter by reference.

Instead of removing GaussianBallWalk::Walk, now we have kept both GaussianBallWalk::Walk and GaussianHamiltonianMonteCarloExactWalk.

Keeping the default parameter GaussianCDHRWalk.

abureyanahmed committed:
No globals in the gaussian HMC

Previously, we were using a global variable, which is a bad practice. Hence, we have removed global variables.

Also, removed unnecessary definitions, headers, and printing.

vissarion commented on include/random_walks/gaussian_hamiltonian_monte_carlo_exact_walk.hpp:
```diff
@@ -57,8 +57,8 @@ struct Walk
     {
-        _Len = compute_diameter<GenericPolytope>
-                ::template compute<NT>(P);
+        //_Len = compute_diameter<GenericPolytope>
```
why are you commenting this out? This is where `_Len` is computed that is a essential parameter of the walk.

abureyanahmed committed:
Added missing function

We have added a function to compute the initial diameter of a polytope.

abureyanahmed committed:
Calculating diameter more appropriately

Previously the diameter was initially set up equal to zero. Now we are setting the diameter more appropriately from the properties of the polytope.

abureyanahmed committed:
Test Gaussian HMC

Since we have a new walk (Gaussian Hamiltonian Monte Carlo exact walk), now we are testing that too.

abureyanahmed committed:
Calculating diameter more appropriately

Previously the diameter was initially set up equal to zero. Now we are setting the diameter more appropriately from the properties of the polytope.

abureyanahmed committed:
No changes are needed here

Previously I had made the modifications in this file unintentionally. This file actually does not need any changes.

abureyanahmed committed:
Passes all tests

Previously one test was failing. Now we have adjusted the accuracy parameter to ensure that it passes all tests.

abureyanahmed committed:
cooling gaussian for general mass matrix

We started the walk at zero. We can initialize in a better way by first estimating the covariance matrix and then using the diagonal entries of that matrix for initialization.

vissarion commented on include/volume/volume_cooling_gaussians_mass_matrix.hpp:
```diff
@@ -0,0 +1,528 @@
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+#ifndef VOLUME_COOLING_GAUSSIANS_HPP
```
please use the whole path as in other places of the library

vissarion commented on include/volume/volume_cooling_gaussians_mass_matrix.hpp:
```diff
@@ -0,0 +1,528 @@
+};
+
+template <typename Polytope, typename RandomNumberGenerator>
+struct update_delta<GaussianHamiltonianMonteCarloExactWalk::Walk<Polytope, RandomNumberGenerator>>
```
is this redefined here https://github.com/GeomScale/volesti/pull/230/files#diff-f34df8328972aebd7b85669d7cf3568caf3720a784967874cde9eb497c9afc23R40  or I am missing something?

vissarion commented on include/volume/volume_cooling_gaussians_mass_matrix.hpp:
```diff
@@ -0,0 +1,528 @@
+
+// Compute the first variance a_0 for the starting gaussian
+template <typename Polytope, typename NT>
+void get_first_gaussian(Polytope const& P,
```
these function as well as all the following are redefined in https://github.com/GeomScale/volesti/blob/develop/include/volume/volume_cooling_gaussians.hpp 
Please try to reuse as much code as possible. This file here should provide a new function `volume_general_cooling_gaussians()` that reuses functions from https://github.com/GeomScale/volesti/blob/develop/include/volume/volume_cooling_gaussians.hpp and some new functions (with the implementation of the general mass matrix).

abureyanahmed committed:
Reusing functions for general mass

In this update, we are reusing as much code as possible from "volume_cooling_gaussians.hpp". And created a new function named volume_general_cooling_gaussians().

abureyanahmed committed:
Merge branch 'GeomScale:develop' into develop

vissarion commented:
Closing this as inactive.
